GitExternal.cmake

```diff
@@ -101,67 +109,68 @@ function(GIT_EXTERNAL DIR REPO TAG)
-      WORKING_DIRECTORY "${DIR}")
-    git_external_message("git checkout -q -- ${GIT_EXTERNAL_RESET_FILE}")
-    execute_process(
-      COMMAND "${GIT_EXECUTABLE}" checkout -q -- "${GIT_EXTERNAL_RESET_FILE}"
-      RESULT_VARIABLE nok ERROR_VARIABLE error
-      WORKING_DIRECTORY "${DIR}")
-  endforeach()
-
-  # fetch latest update
-  execute_process(COMMAND "${GIT_EXECUTABLE}" fetch origin -q
-    RESULT_VARIABLE nok ERROR_VARIABLE error
-    WORKING_DIRECTORY "${DIR}")
-  if(nok)
-    message(STATUS "Update of ${DIR} failed:\n   ${error}")
-  endif()
-
-  # update tag
-  git_external_message("git rebase FETCH_HEAD")
-  execute_process(COMMAND ${GIT_EXECUTABLE} rebase FETCH_HEAD
-    RESULT_VARIABLE RESULT ERROR_QUIET OUTPUT_QUIET
-    WORKING_DIRECTORY "${DIR}")
-  if(RESULT)
-    message(STATUS "git rebase failed, aborting ${DIR} merge")
-    execute_process(COMMAND ${GIT_EXECUTABLE} rebase --abort
-      WORKING_DIRECTORY "${DIR}" ERROR_QUIET OUTPUT_QUIET)
-  endif()
+  file(RELATIVE_PATH __dir ${CMAKE_SOURCE_DIR} ${DIR})
+  string(REGEX REPLACE "[:/]" "-" __target "${__dir}")
+  set(__rebase_cmake "${CMAKE_CURRENT_BINARY_DIR}/${__target}-rebase.cmake")
+  file(WRITE ${__rebase_cmake}
+    "if(NOT IS_DIRECTORY ${DIR}/.git)\n"
+    "  message(FATAL_ERROR \"Can't update git external ${__dir}: Not a git repository\")\n"
+    "endif()\n"
+    # check if we are already on the requested tag
+    "execute_process(COMMAND ${GIT_EXECUTABLE} rev-parse --short HEAD\n"
+    "  OUTPUT_VARIABLE currentref OUTPUT_STRIP_TRAILING_WHITESPACE\n"
+    "  WORKING_DIRECTORY ${DIR})\n"
+    "if(currentref STREQUAL ${TAG}) # nothing to do\n"
+    "  return()\n"
+    "endif()\n"
+    "\n"
+    # reset generated files
+    "foreach(GIT_EXTERNAL_RESET_FILE ${GIT_EXTERNAL_RESET})\n"
+    "  execute_process(\n"
+    "    COMMAND \"${GIT_EXECUTABLE}\" reset -q \"\${GIT_EXTERNAL_RESET_FILE}\"\n"
+    "    ERROR_QUIET OUTPUT_QUIET\n"
+    "    WORKING_DIRECTORY \"${DIR}\")\n"
+    "  execute_process(\n"
+    "    COMMAND \"${GIT_EXECUTABLE}\" checkout -q -- \"${GIT_EXTERNAL_RESET_FILE}\"\n"
+    "    ERROR_QUIET OUTPUT_QUIET\n"
+    "    WORKING_DIRECTORY \"${DIR}\")\n"
+    "endforeach()\n"
+    "\n"
+    # fetch latest update
+    "execute_process(COMMAND \"${GIT_EXECUTABLE}\" fetch origin -q\n"
+    "  RESULT_VARIABLE nok ERROR_VARIABLE error\n"
+    "  WORKING_DIRECTORY \"${DIR}\")\n"
+    "if(nok)\n"
+    "  message(FATAL_ERROR \"Fetch for ${__dir} failed:\n   \${error}\")\n"
+    "endif()\n"
+    "\n"
+    # update tag
+    "execute_process(COMMAND ${GIT_EXECUTABLE} rebase FETCH_HEAD\n"
+    "  RESULT_VARIABLE nok ERROR_VARIABLE error OUTPUT_VARIABLE output\n"
+    "  WORKING_DIRECTORY \"${DIR}\")\n"
+    "if(nok)\n"
+    "  execute_process(COMMAND ${GIT_EXECUTABLE} rebase --abort\n"
+    "    WORKING_DIRECTORY \"${DIR}\" ERROR_QUIET OUTPUT_QUIET)\n"
+    "  message(FATAL_ERROR \"Rebase ${__dir} failed:\n\${output}\${error}\")\n"
```

Comment:
Now that it's an explicit target I'ld like it to fail on failure. One can always do 'ninja rebase -k0'